### PR TITLE
TPCClusterFinder: Add support for >= 4000 time bins.

### PR DESCRIPTION
**Pull request by fweig**
This PR adds support for >= 4000 time bins in the TPCClusterFinder. It also replaces the `PackedDigit` class with `tpc::Digit` to support larger time stamps.

### CODE REVIEW

File: GPU/GPUTracking/Global/GPUChainTracking.cxx [hunk 1]
```diff
@@ -870,171 +870,185 @@ int GPUChainTracking::RunTPCClusterizer()
...
+          }
+        }
+        TransferMemoryResourceLinkToGPU(RecoStep::TPCClusterFinding, clusterer.mMemoryId, lane);
+
+        // These buffers only have to be cleared once entirely. The 'resetMaps' kernel
+        // takes care of subsequent clean ups.
+        if (iSliceBase == 0) {
+          using ChargeMapType = decltype(*clustererShadow.mPchargeMap);
+          using PeakMapType = decltype(*clustererShadow.mPpeakMap);
+          runKernel<GPUMemClean16>({BlockCount(), ThreadCount(), lane, RecoStep::TPCClusterFinding}, krnlRunRangeNone, {}, clustererShadow.mPchargeMap, TPCMapMemoryLayout<ChargeMapType>::items() * sizeof(ChargeMapType));
```
Comment (davidrohr) [inline GPU/GPUTracking/Global/GPUChainTracking.cxx:922]: Could you please stick to using `GetGridBlkStep(BlockCount(), lane, RecoStep::TPCClusterFinding)`

File: GPU/GPUTracking/TPCClusterFinder/clusterFinderDefs.h [hunk 2]
```diff
@@ -94,7 +94,8 @@ namespace GPUCA_NAMESPACE
 namespace gpu
 {
 
-using Timestamp = short;
+using TPCTime = int32_t;
```
Comment (davidrohr) [inline GPU/GPUTracking/TPCClusterFinder/clusterFinderDefs.h:97]: Could you move these definitions into a namespace or class, TPCCF or so, such that they are not globally visible in o2::gpu?

**Comment by davidrohr** [comment]
The following changes are required to make it compile with OpenCL: (OpenCL doesn't not the int32_t etc. Perhaps there is some way to include it, but I just changed it to int for now)
```
diff --git a/DataFormats/common/include/CommonDataFormat/TimeStamp.h b/DataFormats/common/include/CommonDataFormat/TimeStamp.h
index a8e3553d130..fba0cd6a431 100644
--- a/DataFormats/common/include/CommonDataFormat/TimeStamp.h
+++ b/DataFormats/common/include/CommonDataFormat/TimeStamp.h
@@ -49,7 +49,7 @@ class TimeStampWithError : public TimeStamp<T>
   ClassDefNV(TimeStampWithError, 1);
 };
 
-#ifndef ALIGPU_GPUCODE
+#ifndef GPUCA_GPUCODE
 template <typename T>
 std::ostream& operator<<(std::ostream& os, const TimeStamp<T>& t)
 {
diff --git a/GPU/GPUTracking/TPCClusterFinder/clusterFinderDefs.h b/GPU/GPUTracking/TPCClusterFinder/clusterFinderDefs.h
index 9f4840a69ef..105b6ab2a7a 100644
--- a/GPU/GPUTracking/TPCClusterFinder/clusterFinderDefs.h
+++ b/GPU/GPUTracking/TPCClusterFinder/clusterFinderDefs.h
@@ -94,8 +94,8 @@ namespace GPUCA_NAMESPACE
 namespace gpu
 {
 
-using TPCTime = int32_t;
-using TPCFragmentTime = int16_t;
+using TPCTime = int;
+using TPCFragmentTime = short;
 using Pad = unsigned char;
 using GlobalPad = short;
 using Row = unsigned char;
```

**Comment by davidrohr** [comment]
I checked the performance, and it seems the memclean kernel now consumes much more time before, even though the input is the same. Perhaps this is related to the GetBlock thing I commented above?

**Comment by davidrohr** [comment]
@fweig : Is this already supposed to work with ZS data? I have processed some of the 4000 time bin data I have, setting TPC_MAX_TIME to 1000 or 2000, but then I get totally different results.
If I understand correctly, everything is derrived from TPC_MAX_TIME, so there is no need to have the fragment length being 4000 but this is configurable?

**Comment by fweig** [comment]
Yes, this should work with ZS as well. I left TPC_MAX_TIME at 4000 and varied `fragmentLen` in `RunTPCClusterizer` to test different fragment sizes. Tested for 4000, 2000, 1000 and 600 time bins. The produced clusters were identical to `dev` both with and without ZS.

File: GPU/GPUTracking/Global/GPUChainTracking.cxx (hunk 1 above)
Comment (fweig) [reply GPU/GPUTracking/Global/GPUChainTracking.cxx]: Fixed. This also fixes the performance regression.

File: GPU/GPUTracking/TPCClusterFinder/clusterFinderDefs.h (hunk 2 above)
Comment (fweig) [reply GPU/GPUTracking/TPCClusterFinder/clusterFinderDefs.h]: Moved them into namespace `o2::gpu::tpccf`.

**Comment by davidrohr** [comment]
@fweig : I just tried with fragmentLen (line 881 in GPUChainTracking.cxx) but that doesn't work for me.
My input data is zero-suppressed and its max time bin is ~2200.
With TPC_MAX_TIME still at 4000 and fragmentLen = TPC_MAX_TIME, it reconstructs correctly.
With fragmentLen = 1000, I obtain 0 TPC clusters.
I am still not sure about the MAX_TIME constant, is it still the upper limit for the number of time bins, and fragmentLen is number of time bins processed at a time?
And what is the difference between timeSliceLen and fragmentLen? timeSliceLen is set to TPC_MAX_TIME, so if this is the upper limit of number of time bins, we have to raise it to the real TPC time frame length which is 3564 * 256 / 8, so much larger than 4000.
If I understand correctly, also the memory allocation for the chargemap etc scales with TPC_MAX_TIME. Shouldn't this scale with the fragment length now?

Perhaps it is easiest to have a chat about this. Do you have some time tomorrow?

**Comment by fweig** [comment]
Sure, we can talk about this tomorrow. I'm available.

`TPC_MAX_TIME` is poorly named. It represents the maximum number of time bins a fragment can contain. While `timeSliceLen` is the total number of time bins in the data.

**Comment by davidrohr** [comment]
OK, that makes more sense, then we should just rename these a bit. But it still doesn't work for me. I tried reducing TPC_MAX_TIME to 1000 and setting timeSliceLen in GPUChainTracking to 4000, and again I get 0 clusters.

**Comment by davidrohr** [comment]
@fweig : I checked the new version in more detail.
With ZS data it looks all good, but now it seems to be broken when the input is digits.
Could you have a look again?

**Comment by fweig** [comment]
@davidrohr I can't reproduce this. Works with and without ZS for me. Intermediate results differ between the two in the dumped data but the produced clusters are identical.

**Comment by davidrohr** [comment]
@fweig : The following patch might fix your hip problems:

Please note: All GPU()... annotated functions are compiled twice, during host compilation and GPU compilation. You need the GPUhd version only when the GPU management code needs to access the host version of the function compiled by the GPU compiler.

All normal GPU code, e.g. GPUChainTracking, will use the GCC-compiled version of the function, which is anyway there irrespective of GPU???(). So if in doubt, just use GPUd() and remove the h.

Perhaps at some point I should do a general cleanup and change all GPUhd() to GPUd() where not strictly needed.

```
diff --git a/DataFormats/common/include/CommonDataFormat/TimeStamp.h b/DataFormats/common/include/CommonDataFormat/TimeStamp.h
index 148d12e3b3b..c3420d348c6 100644
--- a/DataFormats/common/include/CommonDataFormat/TimeStamp.h
+++ b/DataFormats/common/include/CommonDataFormat/TimeStamp.h
@@ -26,10 +26,10 @@ class TimeStamp
  public:
   GPUdDefault() TimeStamp() CON_DEFAULT;
   GPUdDefault() ~TimeStamp() CON_DEFAULT;
-  GPUhdi() TimeStamp(T time) { mTimeStamp = time; }
-  GPUhdi() T getTimeStamp() const { return mTimeStamp; }
-  GPUhdi() void setTimeStamp(T t) { mTimeStamp = t; }
-  GPUhdi() bool operator==(const TimeStamp<T>& t) const { return mTimeStamp == t.mTimeStamp; }
+  GPUdi() TimeStamp(T time) { mTimeStamp = time; }
+  GPUdi() T getTimeStamp() const { return mTimeStamp; }
+  GPUdi() void setTimeStamp(T t) { mTimeStamp = t; }
+  GPUdi() bool operator==(const TimeStamp<T>& t) const { return mTimeStamp == t.mTimeStamp; }
 
  private:
   T mTimeStamp = 0;
diff --git a/Detectors/TPC/base/include/TPCBase/Digit.h b/Detectors/TPC/base/include/TPCBase/Digit.h
index 0317a6038ad..5756bf037a6 100644
--- a/Detectors/TPC/base/include/TPCBase/Digit.h
+++ b/Detectors/TPC/base/include/TPCBase/Digit.h
@@ -42,7 +42,7 @@ class Digit : public DigitBase
   /// \param charge Accumulated charge of Digit
   /// \param row Global pad row in which the Digit was created
   /// \param pad Pad in which the Digit was created
-  GPUhdi() Digit(int cru, float charge, int row, int pad, int time);
+  GPUdi() Digit(int cru, float charge, int row, int pad, int time);
 
   /// Destructor
   GPUdDefault() ~Digit() CON_DEFAULT;
@@ -50,25 +50,25 @@ class Digit : public DigitBase
   /// Get the accumulated charged of the Digit in ADC counts.
   /// The conversion is such that the decimals are simply stripped
   /// \return charge of the Digit
-  GPUhdi() int getCharge() const { return static_cast<int>(mCharge); }
+  GPUdi() int getCharge() const { return static_cast<int>(mCharge); }
 
   /// Get the accumulated charged of the Digit as a float
   /// \return charge of the Digit as a float
-  GPUhdi() float getChargeFloat() const { return mCharge; }
+  GPUdi() float getChargeFloat() const { return mCharge; }
 
-  GPUhdi() void setCharge(float q) { mCharge = q; }
+  GPUdi() void setCharge(float q) { mCharge = q; }
 
   /// Get the CRU of the Digit
   /// \return CRU of the Digit
-  GPUhdi() int getCRU() const { return mCRU; }
+  GPUdi() int getCRU() const { return mCRU; }
 
   /// Get the global pad row of the Digit
   /// \return Global pad row of the Digit
-  GPUhdi() int getRow() const { return mRow; }
+  GPUdi() int getRow() const { return mRow; }
 
   /// Get the pad of the Digit
   /// \return pad of the Digit
-  GPUhdi() int getPad() const { return mPad; }
+  GPUdi() int getPad() const { return mPad; }
 
  protected:
   float mCharge = 0.f;      ///< ADC value of the Digit
@@ -79,7 +79,7 @@ class Digit : public DigitBase
   ClassDefNV(Digit, 1);
 };
 
-GPUhdi() Digit::Digit(int cru, float charge, int row, int pad, int time)
+GPUdi() Digit::Digit(int cru, float charge, int row, int pad, int time)
   : DigitBase(time),
     mCharge(charge),
     mCRU(cru),
```

**Comment by davidrohr** [comment]
> @davidrohr I can't reproduce this. Works with and without ZS for me. Intermediate results differ between the two in the dumped data but the produced clusters are identical.

OK, that was my fault, I didn't set the maximum time bin correctly.
The current version seems to be fully working.
Once you fix the compilation on HIP, we can merge it.

**Comment by fweig** [comment]
Thanks! HIP patch is applied.

**Pull request merged**